qutebrowser/completion/models/sortfilter.py

```diff
@@ -146,10 +148,44 @@ def filterAcceptsRow(self, row, parent):
                 data = self.srcmodel.data(idx)
                 if not data:
                     continue
-                elif self.pattern.casefold() in data.casefold():
-                    return True
+                else:
+                    matchers = {
+                        'contain': self.containMatcher,
+                        'fuzzy': self.fuzzyMatcher,
+                        'start': self.startMatcher
+                    }
+                    match_type = config.get('completion', 'match-type')
+                    if match_type not in matchers:
+                        match_type = 'contain'
+                    matcher = matchers[match_type]
+                    if matcher(data):
+                        return True
             return False
 
+    def fuzzyMatcher(self, data):
+        """Matcher for 'fuzzy' matching type logic."""
+        data = data.casefold()
+        pattern = self.pattern.casefold()
+        for char in pattern:
+            if char not in data:
+                return False
+        return True
+
+    def containMatcher(self, data):
+        """Matcher for 'contain' matching type logic."""
+        data = data.casefold()
+        pattern = self.pattern.casefold()
+        return pattern in data
+
+    def startMatcher(self, data):
+        """Matcher for 'start' matching type logic."""
+        protocol = re.compile('^(?:(ht|f)tp(s?)://)?')
+        data = data.casefold()
+        pattern = self.pattern.casefold()
+        if protocol.match(data).lastindex is not None:
```

Comment:
I think you could make this simpler by removing the outermost group (`(?:...)?`)  and using `if protocol.match(data) is not None` - or am I missing something?